gax-java/gax/src/main/java/com/google/api/gax/rpc/TransportChannelProvider.java

```diff
@@ -90,6 +90,15 @@ public interface TransportChannelProvider {
   /** True if the TransportProvider has no endpoint set. */
   boolean needsEndpoint();
 
+  /**
+   * TransportChannelProvider allows setting a custom endpoint which may not be the resolved
+   * endpoint. Resolving the endpoint is _always_ required for gRPC and HttpJson Transports.
+   * Unlike {@link #needsEndpoint()}, it is not dependent if the user set a custom endpoint.
+   */
+  default boolean needsResolvedEndpoint() {
+    return false;
+  }
+
```

Comment:
Alternatively, a possible option would be that ClientContext always overrides the endpoint in the TransportChannelProvider (ignoring the `needsEndpoint()` call). This would require changing the behavior of `withEndpoint(...)` for two subclasses: LocalChannelProvider and FixedChannelProvider as they throw an exception on the `withEndpoint(...)` call.

Adding a this new method would prevent requiring changing the behavior of the public classes.